pkg/statedb/types.go

```diff
@@ -56,19 +57,49 @@ type Table[Obj any] interface {
 	// are not possible with a lower bound search.
 	LowerBound(ReadTxn, Query[Obj]) (iter Iterator[Obj], watch <-chan struct{})
 
+	// DeleteTracker creates a new delete tracker for the table.
+	//
+	// It starts tracking deletions performed against the table from the
+	// current revision. A WriteTxn against the target table is required to
+	// add the tracker to the table.
+	DeleteTracker(txn WriteTxn, trackerName string) (*DeleteTracker[Obj], error)
+}
+
+// RWTable provides methods for modifying the table under a write transaction
+// that targets this table.
+type RWTable[Obj any] interface {
+	// RWTable[Obj] is a superset of Table[Obj]. Queries made with a
+	// write transaction return the fresh uncommitted modifications if any.
+	Table[Obj]
+
 	// Insert an object into the table. Returns the object that was
-	// replaced if there was one. Error may be returned if the table
-	// is not locked for writing or if the write transaction has already
-	// been committed or aborted.
+	// replaced if there was one.
+	//
+	// Possible errors:
+	// - ErrTableNotLockedForWriting: table was not locked for writing
+	// - ErrTransactionClosed: the write transaction already committed or aborted
 	//
 	// Each inserted or updated object will be assigned a new unique
 	// revision.
 	Insert(WriteTxn, Obj) (oldObj Obj, hadOld bool, err error)
 
+	// CompareAndSwap compares the existing object's revision against the
+	// given revision and if equal it replaces the object.
+	//
+	// Possible errors:
+	// - ErrRevisionNotEqual: the object has mismatching revision
+	// - ErrObjectNotFound: object not found from the table
+	// - ErrTableNotLockedForWriting: table was not locked for writing
+	// - ErrTransactionClosed: the write transaction already committed or aborted
+	CompareAndSwap(WriteTxn, Revision, Obj) (oldObj Obj, hadOld bool, err error)
+
 	// Delete an object from the table. Returns the object that was
-	// deleted if there was one. Error may be returned if the table
-	// is not locked for writing or if the write transaction has already
-	// been committed or aborted.
+	// deleted if there was one.
+	//
+	// Possible errors:
+	// - ErrObjectNotFound: object not found from the table
```

Comment:
This comment is wrong, `Delete` can't return this error.

Reply:
Fixed.